### PR TITLE
manager: reduce complexity of unit_gc_sweep

### PR DESCRIPTION
**Pull request by lnykryn**
When unit is marked as UNSURE, we are trying to find if it state was
changed over and over again. But only thing that could change is that
some of its dependencies was add to cleanup queue, so it might go to BAD
state, but that does not matter since we remove both UNSURE and BAD
units.

This is a problem on machines with a lot of initscripts with different
starting priority, since those units will reference each other and the
original algorithm might get to n! complexity.

### CODE REVIEW

**Comment by martinpitt** [comment]
CI-wise this is good, the amd64 failure is the logind crash (dev-ttyS0.device hanging) again. Patch is straightforward, but I'd rather let Lennart review this.

**Comment by d-hatayama** [comment]
I think the idea of the patch is good. The n! complexity is reduced to n because unit_gc_sweep() finishs marking all the objects in the 1st call and the 2nd and the latter calls finish soon in constant time. So it may be enough to call unit_gc_sweep() in manager_dispatch_gc_queue() just once. Also, it appears to me that UNSURE is same as BAD in meaning, so another way is to remove UNSURE and replace it with BAD.

**Comment by d-hatayama** [comment]
> because unit_gc_sweep() finishs marking all the objects in the 1st call and the 2nd and the latter calls finish soon in constant time. So it may be enough to call unit_gc_sweep() in manager_dispatch_gc_queue() just once.

Sorry, this was wrong. Two objects in the gc queue could have no dependencies each other in general. Iterating the gc queue is necessary. But complexity is still up to n.

**Comment by poettering** [comment]
Hmm, this doesn't look right to me. Let's say we have a dependency network like this:

```
A←B
B←C
B←D
C←A
```

Where "←" means "referenced by".

We start at A, mark it as "IN_PATH". Go to B, mark it as "IN_PATH", got to C and mark it as "IN_PATH", from there go to A, see it is IN_PATH, go one step back, mark C as "UNSURE",  go one step back (now at B), end go from there to D, which is active, so we mark it GOOD. We go back to B now, which we mark GOOD, too. Then go back to A, mark it GOOD. According to your PR here we wouldn't check C again, because it is marked UNSURE, and we don't look at UNSURE units anymore. After the GC completed we then drop all that are marked BAD and marked UNSURE, and that's C in this case. So we keep around A, B, D. But that's not right, we should have kept C too, as C is referenced by A...

Or did I miss somethin here?

**Comment by lnykryn** [comment]
Yeah, you are right, the patch will not work in this case.
But to the original issue, let's have units ABCD and those reference each other (complete graph) and all of them will be resolved as UNSURE
Then you will go:
A->B->C->D then return to C and B, but now for this point of view path A->B->D->C was not examined. And we will try every possible path in that graph
ACBD,ACDB,ADBC,ADCD.

**Comment by poettering** [comment]
OK, so what about this:

we add a new state CHECK_AGAIN. And then make the following change to the algorithm: we won't look into units with UNSURE state anymore, but we will look into units with CHECK_AGAIN state. And each time we set a unit to BAD or GOOD we will set all its neighbors that are UNSURE to CHECK_AGAIN.

Does that make sense to you?

**Comment by lnykryn** [comment]
I don't think it is enough just mark the neighbours. https://goo.gl/photos/p9Eg8UYtRZPFL85AA If you start with A, go B->C->D->X->E with return you will mark C,D,X,E unsure, got to F (which is good), then return to A and mark AB as GOOD and E,C as CHECK.
If you then start with X, you wil just check D and in is UNSURE. And so you will drop X.

You need to remark almost everything that you marked as UNSURE during that run. We could go through the graph again, but I think that easier solution is to have a special set where we will put all units marked as UNSURE, and we will not revisite unit that is on that list. And if we encounter some GOOD unit, we will just drop the list.

**Comment by d-hatayama** [comment]
There is also UNIT_REFERENCES as an inverse of UNIT_REFERENCED_BY. Is it easier to expand GOOD state via UNIT_REFERENCES when we encounter some GOOD unit?

**Comment by d-hatayama** [comment]
Or, how about expand GOOD state via UNIT_REFERENCES dependency if we encounter some GOOD unit?

**Comment by d-hatayama** [comment]
(Sorry for repeating noise mails again and again...)
Or how about expand GOOD state via UNIT_REFERENCES dependency? Like this:

```
diff --git a/src/core/manager.c b/src/core/manager.c
index bb50503..280e1cd 100644
--- a/src/core/manager.c
+++ b/src/core/manager.c
@@ -765,6 +765,19 @@ enum {
         _GC_OFFSET_MAX
 };

+static void expand_good_state(Unit *u, unsigned gc_marker)
+{
+        Iterator i;
+        Unit *other;
+
+        SET_FOREACH(other, u->dependencies[UNIT_REFERENCES], i) {
+                if (other->gc_marker == gc_marker + GC_OFFSET_UNSURE) {
+                        other->gc_marker = gc_marker + GC_OFFSET_GOOD;
+                        expand_good_state(other);
+                }
+        }
+}
+
 static void unit_gc_sweep(Unit *u, unsigned gc_marker) {
         Iterator i;
         Unit *other;
@@ -774,6 +787,7 @@ static void unit_gc_sweep(Unit *u, unsigned gc_marker) {

         if (u->gc_marker == gc_marker + GC_OFFSET_GOOD ||
             u->gc_marker == gc_marker + GC_OFFSET_BAD ||
+            u->gc_marker == gc_marker + GC_OFFSET_UNSURE ||
             u->gc_marker == gc_marker + GC_OFFSET_IN_PATH)
                 return;

@@ -815,6 +829,7 @@ bad:

 good:
         u->gc_marker = gc_marker + GC_OFFSET_GOOD;
+        expand_good_state(u, gc_marker);
 }

 static unsigned manager_dispatch_gc_queue(Manager *m) {
```

**Comment by lnykryn** [comment]
I came to the same idea and I think that would be the best solution.
Will you create a new PR for that or should I modify this one?

**Comment by lnykryn** [comment]
I have forced-push the branch with the expand good change.

File: src/core/manager.c [hunk 1]
```diff
@@ -877,6 +877,19 @@ enum {
         _GC_OFFSET_MAX
 };
 
+static void unit_gc_mark_good(Unit *u, unsigned gc_marker)
+{
```
Comment (poettering) [inline src/core/manager.c:881]: opening bracket on same line as the function name please, please see CODING_STYLE.

**Comment by poettering** [comment]
ok, makes sense. found only a nitpick. I trust you gave this some substantial testing?

**Comment by lnykryn** [comment]
Works on my machine :-)

**Pull request merged**